### PR TITLE
Playwright V2 E2E Framework

### PR DESCRIPTION
**Pull request by RXminuS**
Completely revamped E2E testing framework for E2E tests, Issue Replications, and faster TDD UI development.

1. eliminates all flake
2. fully isolated tests + fully configurable per-test overrides (vscode version, installed extensions, config, etc)
3. runs faster, fully parallel and headless
4. leverages Playwright tracing UI instead of video recordings
5. allows for easy recording and mocking of network traffic
6. allows for automatic recording of tests by just "using VSCode"
7. "hot-reloadable" extensions & tests (so watch mode works properly)

## Test plan

Not making  part of CI and doesn't effect anyone else yet. Locally we test with `pnpm run test:e2e2`

### CODE REVIEW

**Comment by keegancsmith** [comment]
@RXminuS I just pushed some commits to make this work on my linux machine. Please take a look. Busy reviewing the code as well.

File: vscode/src/local-context/utils.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            // otherwise we assume it's abandoned and we'll give ourselves
+
+            // Note: this could fail if the file has been deleted by another
+            // process right as we check this...I can live with that.
+            const fileStats = await fs.stat(filePath)
+            const age = Date.now() - fileStats.mtimeMs
+            if (age < maxMtimeMs) {
+                // this file has not been abandoned
+                return false
+            }
```
Comment (keegancsmith) [inline vscode/src/local-context/utils.ts:74]: This feels very dodgy. You are trusting that mtime is regularly updated on all filesystems in a reasonably fast way. These are ways I can see this breaking:
- the filesystem just doesn't respect mtime (rare but possible)
- the regularity mtime is updated is more than your maxMtimeMs. eg networked filesystems or possibly filesystems that wait for fsync and have large buffers / slow downloads.
- an attempt to download happened with the wrong clock time in the future and failed. NTP ran and the clock is now correct, leading to us refusing to ever download

I am sure there are more reasons.

It seems like a _lot_ of the complexity added here was to help with e2e tests which can have multiple processes starting up at the same time. Is there a way we can only use the more complicated logic in those cases, and rely on simpler code for the normal user of cody?

File: vscode/src/local-context/utils.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            logDebug('symf', `file ${filePath} is abandoned, removing it`)
```
Comment (keegancsmith) [inline vscode/src/local-context/utils.ts:75]: this helper is also used by bfg now?

**Review by keegancsmith — COMMENTED** [review]
I like this a lot and really want to land it now so we can iterate on it more incrementally. However, I have a concern around the complexity introduced around how we fetch the symf and bfg assets. Is there maybe a way we can approach that differently?

Alright now some other comments:

Where do you expect people to put the tests in the current project hierarchy? Directly in vscode/e2e?

When the fixture fails due to timeout it is really hard to debug. For some reason it doesn't blame to the actual line that was running:

```
  1) [e2e] › e2e/example.test.ts:8:9 › Demonstrations › Show off v2 features ───────────────────────

    Fixture "vscodeUI" timeout of 21126ms exceeded during setup.

       at e2e/utils/vscody/fixture.ts:173

      171 | })
      172 |
    > 173 | const implFixture = _test.extend<TestContext, WorkerContext>({
          |                           ^
      174 |     serverRootDir: [
      175 |         // biome-ignore lint/correctness/noEmptyPattern: <explanation>
      176 |         async ({}, use, testInfo) => {

        at /home/keegan/src/github.com/sourcegraph/cody/vscode/e2e/utils/vscody/fixture.ts:173:27
```

**Comment by RXminuS** [comment]
I think the BFG & Symf asset fetching is a problem in the extension in general, not just for these tests. I first noticed it while trying to fix the original E2E tests. The way to trigger it in VSCode is if for some reason you have two extension hosts and you reload the window with an updated extension it would break as two are concurrently trying to unzip to the same location.

I can definetly clean it up a bit into a cleaner and more isolated abstraction

File: vscode/src/local-context/utils.ts (hunk 1 above)
Comment (RXminuS) [reply vscode/src/local-context/utils.ts:74]: As mentioned below, this is actually not a problem just for the tests, it's a problem (although rare) in the extension in general. But I agree, I wasn't really happy with this either. I'll see if I can think of something simpler (as I did something similar but nicer in some test specific areas)

File: vscode/e2e/README.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+TODO: This will be a nice guide
+
+- `pnpm run test:e2e2:run --ui` to view trace UI locally
```
[the remark this answers is not in the document]
Comment (RXminuS) [reply vscode/e2e/README.md:3]: Yeah I had the same, I'll have a look if I can make that nicer

File: lib/shared/src/utils.ts [hunk 4]
```diff
@@ -99,3 +99,25 @@ export function createSubscriber<T>(): Subscriber<T> {
 export function nextTick() {
     return new Promise(resolve => process.nextTick(resolve))
 }
+
+export type SemverString<Prefix extends string> = `${Prefix}${number}.${number}.${number}`
+
+export namespace SemverString {
+    const splitPrefixRegex = /^(?<prefix>.*)(?<version>\d+\.\d+\.\d+)$/
+    export function forcePrefix<P extends string>(prefix: P, value: string): SemverString<P> {
```
[the remark this answers is not in the document]
Comment (RXminuS) [reply lib/shared/src/utils.ts:107]: This type is used in the two locations where the only "guard" against mis-use was a comment left above it. Especially with the fact that one of them uses a 'v' and the other doesn't made that I had a very hard to debug issue as I simply misread the urls the whole time. This essentially just captures those two original comments as type-checks.

**Comment by RXminuS** [comment]
> When the fixture fails due to timeout it is really hard to debug. For some reason it doesn't blame to the actual line that was running:

I'll add it to the todo list because it's really annoying. I'm not sure if it's just something in Playwright though. You can actually see the original location further up in the logs and in the trace it does show correctly as well

**Comment by keegancsmith** [comment]
Ok the complexity is worth it then. I think it is probably worth avoiding concurrent downloads. I did some research cause I dig systems programming and wondered whats best. Everyone seems to do it differently, sqlite relies on fcntl locks and git relies on lock files but then asks the user to clean up if they are left lying around. Given we are in node, I quite liked the creativity of relying on mkdir as a locking primitive. It also has the same issue around assuming mtime is never in the future, but from reading the code I don't think it has a lot of the other potential pitfalls. https://www.npmjs.com/package/proper-lockfile

File: vscode/src/graph/bfg/download-bfg.ts [hunk 5]
```diff
@@ -1,146 +1,189 @@
...
-        const bfgFilename = `cody-engine-${bfgVersion}-${platform}-${rfc795Arch}`
-        const bfgPath = path.join(bfgContainingDir, bfgFilename)
-        const isAlreadyDownloaded = await fileExists(bfgPath)
-        if (isAlreadyDownloaded) {
-            logDebug('CodyEngine', `using downloaded path "${bfgPath}"`)
-            return bfgPath
-        }
+// this protects agains multiple async functions in the same node process from
+// starting a download
+const processDownloadLock = new Mutex()
```
[the remark this answers is not in the document]
Comment (RXminuS) [reply vscode/src/graph/bfg/download-bfg.ts:55]: I don't really have a preference, I just didn't think it  worth having two entirely different methods in what is functionally the same code between BFG and SYMF downloads. I thought it looked more explicit with the Mutex what the aim was.

File: vscode/e2e/utils/vscody/fixture.ts [hunk 6]
```diff
@@ -466,8 +472,34 @@ const implFixture = _test.extend<TestContext, WorkerContext>({
             // so we need to do some magic to get the actual port.
             // TODO: this might not be very cross-platform
             const port = await getPortForPid(codeProcess.pid)
-
             const config = { url: `http://127.0.0.1:${port}/`, token: token }
+
+            // we now need to wait for the server to be downloaded
```
[the remark this answers is not in the document]
Comment (RXminuS) [reply vscode/e2e/utils/vscody/fixture.ts:477]: Doing multiple version handling and switching I will get to later. Current setup aligns with how we already electron download latest version.

[the remark this answers is not in the document]
Comment (RXminuS) [reply vscode/e2e/utils/vscody/fixture.ts:477]: Reason for not doing global hook is I want to be able to on a test by test basis switch to specific vscode commit if I want to (especially for issue reproductions)

**Comment by keegancsmith** [comment]
@RXminuS just tried this out on linux, it is failing. I don't think the waiting for server stuff is working. I can look into it tomorrow. 

```
2) [e2e] › e2e/example.test.ts:8:9 › Demonstrations › Show off v2 features › Start VSCode Session 

    TimeoutError: locator.waitFor: Timeout 10000ms exceeded.
    Call log:
      - waiting for locator('iframe.web-worker-ext-host-iframe')


       at e2e/utils/vscody/uix/vscode.ts:165

      163 |
      164 |         // wait for the UI to be ready
    > 165 |         await page.locator('iframe.web-worker-ext-host-iframe').waitFor({
          |                                                                 ^
      166 |             state: 'attached',
      167 |             timeout: 10000,
      168 |         })

        at /home/keegan/src/github.com/sourcegraph/cody/vscode/e2e/utils/vscody/uix/vscode.ts:165:65
        at /home/keegan/src/github.com/sourcegraph/cody/vscode/e2e/example.test.ts:29:9
```

Another interesting thing I am noticing is this output, I don't know if we should be adding these sort of node specific APIs to cody/shared?

```
[0] All files were successful downloaded, check resources/wasm directory
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "node:path" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/lib/shared/src/lockfile.ts". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[0] 
[0]   dist/post-uninstall.js      998.7kb
[0]   dist/post-uninstall.js.map    2.1mb
[0] 
[0] ⚡ Done in 195ms
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "fs" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/node_modules/.pnpm/graceful-fs@4.2.11/node_modules/graceful-fs/graceful-fs.js". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "util" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/node_modules/.pnpm/graceful-fs@4.2.11/node_modules/graceful-fs/graceful-fs.js". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "assert" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/node_modules/.pnpm/graceful-fs@4.2.11/node_modules/graceful-fs/graceful-fs.js". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[0] 
[0]   dist/extension.node.js      15.0mb ⚠️
[0]   dist/extension.node.js.map  23.5mb
[0] 
[0] ⚡ Done in 386ms
[0] pnpm run -s _build:esbuild:desktop exited with code 0
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "stream" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/node_modules/.pnpm/graceful-fs@4.2.11/node_modules/graceful-fs/legacy-streams.js". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[1] [plugin:vite:resolve] [plugin vite:resolve] Module "constants" has been externalized for browser compatibility, imported by "/home/keegan/src/github.com/sourcegraph/cody/node_modules/.pnpm/graceful-fs@4.2.11/node_modules/graceful-fs/polyfills.js". See https://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[1] pnpm run -s _build:webviews --mode development exited with code 0
```

**Comment by RXminuS** [comment]
> Just tried this out on linux, it is failing. I don't think the waiting for server stuff is working. I can look into it tomorrow.

I forgot to await the downloading 🤦

File: vscode/e2e/utils/vscody/fixture.ts [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+                        throw new Error(`Could not find a vscode executable in ${cliExecutableDir}`)
+                    }
+                    return path.join(cliExecutableDir, vscodeExecutableName)
+                })
+            ).catch(async () => {
+                throw new Error(
+                    `Could not find a vscode executable under ${path.dirname(electronExecutable)}`
+                )
+            })
+            const [cliPath, ...cliArgs] = resolveCliArgsFromVSCodeExecutablePath(vscodeExecutable)
```
Comment (keegancsmith) [inline vscode/e2e/utils/vscody/fixture.ts:382]: this doesn't work right now on linux due to the fact that vscodeExecutable is pointing to code-tunnel.exe not the code executable. I will push commits very soon if I get it to work.

File: vscode/e2e/utils/vscody/fixture.ts [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+                `--cli-data-dir=${serverExecutableDir.replace(/ /g, '\\ ')}`,
+                `--server-data-dir=${serverRootDir.replace(/ /g, '\\ ')}`,
+                `--extensions-dir=${extensionsDir.replace(/ /g, '\\ ')}`, // cli doesn't handle quotes properly so just escape spaces,
+                ...cliArgs,
+            ]
+            //TODO(rnauta): better typing
+            const env = {
+                // inherit environment
+                // TODO: Check why this was necessary. Shouldn't be needed
+                // ...process.env,
```
[the remark this answers is not in the document]
Comment (RXminuS) [reply vscode/e2e/utils/vscody/fixture.ts:472]: Ah sorry please ignore these; I more meant from the perspective of what specifically vscode was lacking given I thought it was running in portable mode.

But I've figured it out now. Will add this in again in a bit

**Comment by RXminuS** [comment]
I have a commit coming that fixes some of this. Just tested on Mac and Linux. Just doing a windows run now

**Comment by RXminuS** [comment]
There was a whole bunch of other funk going on w.r.t. extensions

**Comment by keegancsmith** [comment]
Latest commit works for me on linux if I just do `pnpm run test:e2e2:run`!

FYI the install deps thing fails since that only supports ubuntu which I don't use. I can live with that since CI uses ubuntu.

**Comment by keegancsmith** [comment]
On windows this is the value for me for `sharedExtensionDir` is `C:\Users\keegan\.vscode-server\extensions`. I'm assuming we can use this based on the comments, so you just need to do 

```diff
-                if (!sharedExtensionsDir.endsWith('.vscode-server/extensions')) {
+                if (
+                    !(
+                        sharedExtensionsDir.endsWith('.vscode-server/extensions') ||
+                        sharedExtensionsDir.endsWith('.vscode-server\\extensions')
+                    )
+                ) {
```

With that if I run one test at a time it works! However, when doing parallel runs I get this error

```
Error: EBUSY: resource busy or locked, unlink 'C:\Users\keegan\Code\cody\.test\runs\01J1T6FA83FRCKZYNG1B6H9Y7E\test-vscode-server-XXXXXX7U4YEC\data\logs\20240702T190446\network.log'
    at unlinkSync (node:fs:1877:11)
    at _unlinkSync (node:internal/fs/rimraf:214:14)
    at rimrafSync (node:internal/fs/rimraf:195:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at node:internal/fs/rimraf:253:9
    at Array.forEach (<anonymous>)
    at _rmdirSync (node:internal/fs/rimraf:250:7)
    at rimrafSync (node:internal/fs/rimraf:193:7)
    at rmSync (node:fs:1264:10)
    at Object.<anonymous> (C:\Users\keegan\Code\cody\vscode\playwright.v2.config.ts:22:15)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module.f._compile (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\utilsBundleImpl.js:16:994)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Object.i.<computed>.ut._extensions.<computed> (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\utilsBundleImpl.js:16:1010)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Function.Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at requireOrImport (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\transform\transform.js:192:20)
    at loadUserConfig (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\common\configLoader.js:96:83)
    at loadConfig (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\common\configLoader.js:101:28)
    at loadConfigFromFileRestartIfNeeded (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\common\configLoader.js:258:16)
    at runTests (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\program.js:191:76)
    at t.<anonymous> (C:\Users\keegan\Code\cody\node_modules\.pnpm\playwright@1.44.1\node_modules\playwright\lib\program.js:54:7) {
  errno: -4082,
  code: 'EBUSY',
  syscall: 'unlink',
  path: 'C:\\Users\\keegan\\Code\\cody\\.test\\runs\\01J1T6FA83FRCKZYNG1B6H9Y7E\\test-vscode-server-XXXXXX7U4YEC\\data\\logs\\20240702T190446\\network.log'
}
```

**Comment by keegancsmith** [comment]
The thing I mentioned earlier of not shutting down the process properly is the root cause of the EBUSY. Basically the process just keeps on running even once playwright is done. On windows you can't delete a file if a process still has it open (which we do on startup to clean up old runs). I just checked who had network.log open and it was this process

image

**Comment by RXminuS** [comment]
Yeah I was trying to figure that out. It seems that on windows orphaned processes are made root processes instead of being killed. There's a helper function in the vscode-server repo to work around this that we'll need to copy.

**Review by keegancsmith — APPROVED** [review]
We are gonna land this now without the CI changes. So this code is not 100% ready for others, but it won't effect anyone and will allow us to work on this incrementally from this base.

**Pull request merged**